### PR TITLE
[WIP] Check for overflow on integer arithmetic

### PR DESCRIPTION
**Pull request by chriseth**
related to https://github.com/ethereum/solidity/issues/2274

This is, mostly, a starting place and a very rough and very early
work-in-progress for detecting possible arithmetic overflow conditions.
Right now, it works for basic arithmetic and assignment, does not
analyze branches or control structures, and probably works only for
one function in one contract.

### CODE REVIEW

**Comment by chriseth** [comment]
Thanks for the pull request!

I think that in general, it would probably make sense to separate the "range" and the "allowed range" aspect, also looking towards a more generalized use of these classes. Furthermore, what happens in a binary operator (unless it is a comparison operator) is that the operands are first converted into the common type and then the operation is performed in that common type. Since these are implicit conversions, the range will always fit, but the overflow will happen in the new type. Also, I think it would be beneficial to have a bounds constructor taking a type object (instead of "number of bytes" and "signed"). This means, the operand visitor would be something like:

```
Bounds b = combineBounds(operand, leftBounds, rightBounds);
if (!b.fits(Bounds(commonType)))
  warning(...).
```

Ah and also note that variables can be used before they are declared, so this basically means that whenever `m_varOverflowInfo` does not contain `&var`, just initialize it with the default range of `var`'s type and you do not have to care about `visit(VariableDeclaration...)` at all.

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -215,6 +216,8 @@ bool CompilerStack::analyze()
 				{
 					contract->setDevDocumentation(Natspec::devDocumentation(*contract));
 					contract->setUserDocumentation(Natspec::userDocumentation(*contract));
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: This is not strictly related to this PR, but why is the documentation stuff in this conditional?

File: libdevcore/Common.h [hunk 2]
```diff
@@ -101,6 +101,16 @@ using u160s = std::vector<u160>;
 using u256Set = std::set<u256>;
 using u160Set = std::set<u160>;
 
+// Numeric types with checked arithmetic
+using safe_u64 =  boost::multiprecision::number<boost::multiprecision::cpp_int_backend<64, 64, boost::multiprecision::unsigned_magnitude, boost::multiprecision::checked, void>>;
```
Comment (chriseth) [inline libdevcore/Common.h]: These are not really required, are they?

File: libsolidity/analysis/OverflowBounds.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include <libsolidity/parsing/Token.h>
+#include <libdevcore/Common.h>
+#include <limits>
+
+namespace dev
+{
+namespace solidity
+{
+
+class OverflowBounds
```
Comment (chriseth) [inline libsolidity/analysis/OverflowBounds.h]: Is the word "Overflow" justified here? What about something like "ValueRanges" (we might generalize this to something not just about bounds).

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+		OverflowBounds const& left,
+		OverflowBounds const& right
+		)
+{
+	auto const lbounds = dynamic_cast<IntegerOverflowBounds const*>(&left);
+	auto const rbounds = dynamic_cast<IntegerOverflowBounds const*>(&right);
+	solAssert((lbounds && rbounds), "Passed non-integer bounds objects to binary op for an integer bounds");
+	performBinaryOp(op, *lbounds, *rbounds);
+}
+	void IntegerOverflowBounds::performBinaryOp(
```
Comment (chriseth) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: I somehow expected this function to be static and returning a new `IntegerOverflowBounds`. Why is that not the case?

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+{
+	bool isSigned = type.isSigned();
+	uint16_t bytes = type.storageBytes();
+	uint32_t exponent = bytes * (isSigned ? ((bytes * 8) - 1) : (bytes * 8));
+	dev::bigint allowedMin;
+	dev::bigint allowedMax;
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
```
Comment (chriseth) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: Perhaps there should be a separate file that has free functions that return the min / max values for each type?

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+		allowedMax = allowedMax - 1;
+	}
+	else
+	{
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	return (min >= allowedMin && max <= allowedMax);
```
Comment (chriseth) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: You could use something like `isIncludedIn(IntegerOverflowBounds::maximumRange(type))`

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+	max = allowedMax;
+}
+
+void IntegerOverflowBounds::setToMaximumRange(RationalNumberType const& type)
+{
+	min = numerator(type.value());
+	max = numerator(type.value());
+}
+
+void IntegerOverflowBounds::setToDefaultValue()
```
Comment (chriseth) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: I think it would be cleaner if this class was immutable and these functions would be constructors or factory functions. What do you think?

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+{
+	min = 0;
+	max = 0;
+}
+
+void IntegerOverflowBounds::constrain(OverflowBounds const& bounds)
```
Comment (chriseth) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: Perhaps it would be clearer to just call it `intersect`?

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+{
+	auto const integerBounds = dynamic_cast<IntegerOverflowBounds const*>(&bounds);
+	solAssert(integerBounds, "Only accepting assignment from Integer and Rational");
+	operator=(*integerBounds);
+	return *this;
+}
+
+OverflowBounds& IntegerOverflowBounds::operator=(RationalNumberType const& number)
+{
+	min = numerator(number.value());
```
Comment (chriseth) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: Please handle the case (assertion could be fine for now) where `denominator(number) != 1`

File: libsolidity/analysis/OverflowChecker.cpp [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+		_contract.accept(*this);
+	}
+	catch (FatalError const&)
+	{
+		// We got a fatal error which required to stop further checking, but we can
+		// continue normally from here.
+		if (m_errorReporter.errors().empty())
+			throw; // Something is weird here, rather throw again.
+	}
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: This component should only ever produce warnings, right? Perhaps it does not make sense to return a bool from this function anyway.

File: libsolidity/analysis/OverflowChecker.cpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+bool OverflowChecker::visit(Assignment const& _assignment)
+{
+	_assignment.rightHandSide().accept(*this);
+	auto const identifier = dynamic_cast<Identifier const*>(&_assignment.leftHandSide());
+	if (identifier)
+	{
+		shared_ptr<OverflowBounds>bounds = m_overflowInfo[&_assignment.rightHandSide()];
+		auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: Please always assert after such a `dynamic_cast`.

File: libsolidity/analysis/OverflowChecker.cpp [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+		m_varOverflowInfo[var] = bounds;
+		if (!bounds->fitsType(*(var->annotation().type)))
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: If there is such a more or less symmetric if-else, I would suggest to use the non-negated case first.

File: libsolidity/analysis/OverflowChecker.cpp [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+{
+	lockConstantExpression(_literal);
+	return true;
+}
+
+bool OverflowChecker::visit(BinaryOperation const& _operation)
+{
+	if (lockConstantExpression(_operation))
+		return false; // Do not descend
+	else
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: To reduce indentation, I would suggest to not use an explicit `else` case here.

File: libsolidity/analysis/OverflowChecker.cpp [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+	{
+		// Need to do more work...
+		auto const* integerType = dynamic_cast<IntegerType const*>(_operation.annotation().commonType.get());
+		if (integerType)
+		{
+			//First, descend and resolve both sides of the operation.
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: I think we should do that in any case, regardless of the type. Consider:

`f(a+1)` - if `f` returns `void`, we still want to look deeper into the ast.

Comment (chriseth) [reply libsolidity/analysis/OverflowChecker.cpp]: Actually it might make sense to use `endVisit` for the regular checks and only `visit` for the `lockConstantExpression` calls.

File: libsolidity/analysis/OverflowChecker.cpp [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+
+			IntegerOverflowBounds bounds(*integerType);
+			shared_ptr<OverflowBounds> leftBounds = m_overflowInfo[&(_operation.leftExpression())];
+			shared_ptr<OverflowBounds> rightBounds = m_overflowInfo[&(_operation.rightExpression())];
+
+			bounds.performBinaryOp(_operation.getOperator(), *leftBounds, *rightBounds);
+			m_overflowInfo[&_operation] = make_shared<IntegerOverflowBounds>(bounds);
+			if (!bounds.fitsType(*integerType))
+				m_errorReporter.warning(_operation.location(), "Expression potentially overflowing its storage.  Range is: " + bounds.toString() + " Storage is: "+integerType->toString(false));
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: Please don`t use the word storage here, it has a special meaning in Solidity :-)

File: libsolidity/analysis/OverflowChecker.cpp [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+			//The most naive form of this will only work if the LHS is an identifier.
+			if (Token::isCompareOp(_operation.getOperator()))
+			{
+				auto const identifier = dynamic_cast<Identifier const*>(&_operation.leftExpression());
+				if (identifier)
+				{
+					auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+					if (var)
+					{
+						IntegerOverflowBounds constraint(*_operation.leftExpression().annotation().type);
```
Comment (chriseth) [inline libsolidity/analysis/OverflowChecker.cpp]: In the long run, the constraints should have different mechanics than the ranges as outlined in https://github.com/ethereum/solidity/issues/2274#issuecomment-310323184 - perhaps there should even be a different class that applies them.

**Review by chriseth — COMMENTED** [review]
Would be great to already have tests for cases that work to see the progress :-)

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	else
+	{
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
```
Comment (axic) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: Would these be easier to read as `(1 << exponent) - 1`?

Comment (axic) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: Also this other change (https://github.com/ethereum/solidity/pull/2538#discussion_r126530260) could be utilized here.

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 18]
```diff
@@ -0,0 +1,267 @@
+#include <libsolidity/analysis/IntegerOverflowBounds.h>
```
Comment (axic) [inline libsolidity/analysis/IntegerOverflowBounds.cpp:22]: Missing license header.

File: libsolidity/ast/ASTAnnotations.h [hunk 19]
```diff
@@ -23,6 +23,7 @@
 #pragma once
 
 #include <libsolidity/ast/ASTForward.h>
+#include <libsolidity/analysis/OverflowBounds.h>
```
Comment (axic) [inline libsolidity/ast/ASTAnnotations.h]: I think this is not needed?

**Comment by axic** [comment]
Please rebase.

File: libdevcore/Common.h (hunk 2 above)
Comment (roadriverrail) [reply libdevcore/Common.h]: No.  That's dead code.  I'm removing it in a cleanup.

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 4 above)
Comment (roadriverrail) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: There's no strong reason for it.  It'll probably make for cleaner code in its consumer.  Will rewrite.

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 5 above)
Comment (roadriverrail) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: I'm not against it.  I want to do a code reorg before I proceed.  When you say "each type," do you mean IntegerType and RationalNumberType?

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 17 above)
Comment (roadriverrail) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: I see that #2554 has been merged, so I can rebase to pick that up and then get rid of this redundant code.  Thanks for doing that!

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 7 above)
Comment (roadriverrail) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: I think you're probably on the right track.  Throughout Overflow checker, I'm making "blank" range objects, then pretty much immediately doing something to manipulate them.  After checking in the if/else code earlier today, I was starting to feel like maybe I needed to have functions that return range objects after manipulating them.

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 8 above)
Comment (roadriverrail) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: Yes, with the proviso that I'm calling it `intersectBounds` to match `unionBounds`, because `union` is a reserved word.

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 9 above)
Comment (roadriverrail) [reply libsolidity/analysis/IntegerOverflowBounds.cpp]: Will fix.

File: libsolidity/analysis/OverflowBounds.h (hunk 3 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowBounds.h]: No, probably not.  I had wanted to make a generic class that would be useful for both reasoning about integers and booleans.  The more I work with it, however, the worse that idea even seems.  I am starting to feel like I should just get rid of the parent class until I know what I'm doing with it anyway.  What do you think?

File: libsolidity/analysis/OverflowChecker.cpp (hunk 10 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: Good point.  Easy fix.

File: libsolidity/analysis/OverflowChecker.cpp (hunk 11 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: This seems like a good place to ask the question.  I feel like I'm making a lot of `dynamic_casts` and I feel like it might be a sign of awkward code.  Here, it's obviously needed, but elsewhere, I'm often doing it just to force type compliance, and something feels wrong about that.

File: libsolidity/analysis/OverflowChecker.cpp (hunk 12 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: So, you would prefer:
```
                if (bounds->fitsType(*(var->annotation().type))) { /* do nothing */}
                else
                        m_errorReporter.warning(_assignment.location(), "Variable assignment may cause overflow.  " + bounds->toString());
```

File: libsolidity/analysis/OverflowChecker.cpp (hunk 13 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: Good point.

File: libsolidity/analysis/OverflowChecker.cpp (hunk 14 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: Right now, my code can't support function calls.  When I do get to that point, I will definitely *have* to evaluate the expressions in the parameters to a call because a call like `f(x++)` will generate a side-effect on `x`.  So, I will definitely descend all function calls.  In your example, when that descent reaches `a+1`, I wouldn't expect `lockConstantExpression( )` to filter it out because the expression `a+1` can't be a RationalNumberType.

File: libsolidity/analysis/OverflowChecker.cpp (hunk 16 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: I agree that constraints require a redesign, and I actually think that we'll ultimately need some sort of worker class that can generate them.  I'm just not 100% sure how I want that to work right now.

Really, ranges need a redesign, too, because they can't support being disjoint, but I'm not sure how I want that to work either (possibly with range objects being able to be composed of multiple ranges, too).

File: libsolidity/analysis/OverflowChecker.cpp (hunk 15 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: Will fix.

File: libsolidity/ast/ASTAnnotations.h (hunk 19 above)
Comment (roadriverrail) [reply libsolidity/ast/ASTAnnotations.h]: Will fix.

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (roadriverrail) [reply libsolidity/interface/CompilerStack.cpp]: I don't know the first thing about documentation, so I am not qualified to answer your question.

**Review by roadriverrail — COMMENTED** [review]
@chriseth @axic Thank you both for the attention to this change.  This feature still has a very long way to go, but I appreciate the feedback deeply so that I can fix a bunch of things as a I go.  I will be pushing a commit with your smaller corrections tonight.  I want to start looking at a refactor next so I have cleaner code at the top level.  I might take a break to write some tests first, though.

File: libsolidity/analysis/IntegerOverflowBounds.cpp (hunk 18 above)
Comment (axic) [reply libsolidity/analysis/IntegerOverflowBounds.cpp:22]: Fixed.

File: libsolidity/ast/Types.cpp [hunk 20]
```diff
@@ -414,6 +414,14 @@ u256 IntegerType::literalValue(Literal const* _literal) const
 	return u256(_literal->value());
 }
 
+s256 IntegerType::signedValue(Literal const* _literal) const
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: This doesn't seem to be used yet?

I wonder if `rational value()` would be a better choice since it would be a unified interface between this and `RationalNumberType`?

File: libsolidity/analysis/OverflowChecker.cpp (hunk 14 above)
Comment (roadriverrail) [reply libsolidity/analysis/OverflowChecker.cpp]: Actually, I need to correct myself here.  Right now, in something like `f(a+1)`, we will check the `a+1` because it's a BinaryOperation.

**Comment by roadriverrail** [comment]
Rebased and moved on to the helper functions for IntegerType.  Next on my list is to look to more of the design and refactoring remarks from @chriseth because the code is definitely becoming unwieldy to work with.

Other things I feel are coming up on the horizon that I need:

- Something to properly evaluate boolean expressions and determine the set of constraints that can be applied to ensure that it is true or false.  E.g. given the expression (x || y), it returns {{x}, {y}, {x,y}}.  This will be necessary for properly analyzing if/else, require( ), etc.

- Loops.  I have no good ideas about how to handle them.  The potential for overflow with a loop depends strongly on how many times the loop will run before terminating, and I don't have good intuitions about how to handle this.

That said, I do have some sense about how to do some refactoring to get this code cleaner and easier to read, and I also can work on extending this to analyze more than one function, and I can also work on expanding the mathematical operations supported, and so on.

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+		min = right.min;
+		max = right.max;
+	}
+}
+
+void IntegerOverflowBounds::swapIfNecessary()
+{
+	if (min > max)
+	{
+		dev::bigint swap = min;
```
Comment (axic) [inline libsolidity/analysis/IntegerOverflowBounds.cpp]: Could this use boost's `swap` or is that overkill?

**Comment by chriseth** [comment]
@roadriverrail I have not read all the recent changes, just commenting on your comment above:

I think a simple first version does not require the evaluation of all boolean expression perhaps apart from either simple `&&` or `||`, whichever is more convenient in that situation.

Concerning loops: What you can do is

 - analyze the loop about which local variables it might be affected
 - assume the loop condition inside the body to be true (just like an `if`), but clear all other assumptions about affected local variables
 - after the loop, clear all assumptions about affected local variables (even the loop codition).

**Comment by roadriverrail** [comment]
@chriseth No worries about all the other comments.  I think your
refactoring to make the range objects immutable is a great idea, and I hope
to take it on either later today or tomorrow morning.  Can you explain what
you mean when you say "clear assumptions"?  If I "clear assumptions" from a
particular value range, am I basically returning it to [minval, maxval]?

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+	}
+	else if (_op == Token::GreaterThan)
+	{
+		min = _right.getMax()+1;
+	}
+	else if (_op == Token::Equal)
+	{
+		min = _right.getMin();
+		max = _right.getMax();
+	}
```
Comment (axic) [inline libsolidity/analysis/IntegerOverflowBounds.cpp:249]: Should be an assert for the else here.

File: libsolidity/analysis/IntegerOverflowBounds.cpp [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+		Token::Value const& _op,
+		IntegerOverflowBounds const& _left,
+		IntegerOverflowBounds const& _right
+		)
+{
+	dev::bigint min;
+	dev::bigint max;
+	// see here: https://en.wikipedia.org/wiki/Interval_arithmetic
+	// Not sure if modulus is right, but it feels right
+	if (_op == Token::Add)
```
Comment (axic) [inline libsolidity/analysis/IntegerOverflowBounds.cpp:153]: Would a switch make more sense here?

File: libsolidity/analysis/OverflowChecker.cpp [hunk 24]
```diff
@@ -0,0 +1,301 @@
...
+							invertOp = Token::GreaterThan;
+							break;
+						case Token::GreaterThan:
+							invertOp = Token::LessThan;
+							break;
+						case Token::Equal:
+							invertOp = Token::Equal;
+							break;
+						default:
+							break;
```
Comment (axic) [inline libsolidity/analysis/OverflowChecker.cpp:153]: Assert for this case?

**Comment by roadriverrail** [comment]
I've gone back to trying to consider loops again.  What I'm not really understanding in the explanation of loop analysis provided by @chriseth is how we determine when a loop will or will not produce an overflow.  It might be easier to work through an example of what I consider to be a short but logically challenging loop:

```
uint8 x = 0;
uint8 y = 0;
while (x < some_array.length) {
  y += x;
  x++;
}
```

This loop will overflow `y` for `some_array.length` > 22, but in order for me (personally) to calculate that, I had to recognize that `y` represents the triangular number sequence and to then apply the formula for triangular numbers for the value 255 and arrive at the result, 22, as being the largest triangular number <= 255.  I'm sure there's a way to algorithmically analyze this, but I don't know where to get started to even find out.  It requires recognizing the rates of change of `x` and `y` inside the loop and then working out how many times the loop can execute before it will overflow.  Determining rates of change that are constant across loop iterations (e.g. `x++`) is not so hard because you can simply compare the values at the beginning and end of the loop.  But it's determining rates of change that will vary with each loop iteration (e.g. `y += x` when `x` also changes in the loop) that I can't figure out.

There's also the question of how to determine if the logic of the loop will ensure it runs only a certain number of times.  I'm not clear on how, in a general case, we would be able to take "this loop is safe for 23 iterations" and apply it back to the constraints in the `while( )` expression.

It's possible I don't understand the scope envisioned here or that I don't know of a good method for solving this issue because it's outside my usual area of expertise.  @chriseth, what do you think?

**Comment by axic** [comment]
Can you please rebase and use the new experimental feature pragma?

**Comment by axic** [comment]
@chriseth @leonardoalt @ekpyron @bit-shift @christianparpart do you think it would be worth the effort finishing this PR or all we in favour of dropping this for the SMT option we have? The main pain point with SMT solvers we have is their code size and as a result we cannot bundle it in the compiler. This PR was started around the same time as the SMT prototype work and after that showed signs of success, the idea here was mostly to have a crude overflow detection system.

**Comment by leonardoalt** [comment]
I lean toward the SMT option, assuming we manage to finish the JS work.

Related to that we have the runtime overflow checks we've been talking about, still without conclusive experimental results.

**Comment by axic** [comment]
I tend to be on the same opinion that runtime overflow checks should provide the safeguard (and thus be detected during testing), while SMT would give the tool to help identify those earlier in development.

Downside is that if a runtime test is missed, the contract may end up in an unrecoverable state, but that is just a result of inadequate testing and auditing.

**Comment by ekpyron** [comment]
Working towards making this PR reliable in all or at least most cases would probably mean reimplementing a custom SMT solver anyways, so I'd also lean towards dropping this in favor of the SMT option.
